IDMapper map ids by native_id

RobertDeibel opened the pull request:
# Description
Implement IDMapper mapping by native_id. Rearranging code, and adding native_id retrieval and mapping to spectrum. 
Closes #4965

# Checklist:
- [ ] Make sure that you are listed in the AUTHORS file
- [ ] Add relevant changes and new features to the CHANGELOG file
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] New and existing unit tests pass locally with my changes
- [x] Updated or added python bindings for changed or new classes. (Tick if no updates were necessary.)

Note: If your PR is failing you can check out http://cdash.openms.de/index.php?project=OpenMS and look for your PR with detailed error messages.

RobertDeibel commented:
I'm unsure if the tests cover this. How would one go about writing test for this issue?

jpfeuffer commented on src/openms/source/ANALYSIS/ID/IDMapper.cpp:
```diff
@@ -115,84 +118,107 @@ namespace OpenMS
-      {
-        identifications_precursors.insert(make_pair(peptide_ids[i].getRT(), i));
+      { // mapping is done by either native id or by comparing peptide_id RT with experiment RT
+        if (peptide_ids[i].metaValueExists("spectrum_reference") == false) 
```
```suggestion
        if (!peptide_ids[i].metaValueExists("spectrum_reference")) 
```

jpfeuffer commented on src/openms/source/ANALYSIS/ID/IDMapper.cpp:
```diff
@@ -115,84 +118,107 @@ namespace OpenMS
+    lookup.readSpectra(map);
+
+    // remember which peptides were mapped (for stats later)
+    set<Size> peptides_mapped;
```
Maybe use unordered set if the order does not matter.

jpfeuffer reviewed (changes requested):
For tests, you can copy the input files from the old tests, add some "spectrum_reference" metavalues to the PeptideIdentifications and add an entry after the old one in the tests/topp/CmakeLists.txt

timosachsenberg commented on src/openms/source/ANALYSIS/ID/IDMapper.cpp:
```diff
@@ -115,84 +118,107 @@ namespace OpenMS
-    // once we finish a spectrum, we simply move back in the peptide id window a little to get from the
-    // right end of the old interval to the left end of the new interval
-    while (experiment_iterator != experiment_precursors.end())
+    if (identifications_precursors.size() > 0) 
```
```suggestion
    if (!identifications_precursors.empty()) 
```

RobertDeibel commented:
I tried to add tests as you suggested @jpfeuffer, just copied test 1 and added some spectrum_reference meta values but I'm still not sure how the tests are actually performed. As in, I intentionally tried breaking the `IDMapper` class and while the `IDMapper_test` of tests/openms failed, all the TOPP tests still passed.

I'm obviously missing something but I'm not sure how I should go about on understanding the tests.

jpfeuffer commented:
Yes, so we have two types of tests:
- Class tests
- and TOPP/Tool tests

Class tests are standalone "additional" executables that include the class to be tested and our own testing utility classes to test outcomes of single functions of the class in question.

Tool tests are using the tool executable that the user would also receive. We use those executables to run the full algorithm on a small test dataset, to ensure that from version to version the results stay the same. The fact that the TOPP test did still pass is a bit weird, since it tests the complete algorithm. Maybe you did not break it enough. Often TOPP tests only use standard settings. Alternative "modes" are unfortunately not always tested. In your case, however, your new alternative mode needs to be added now.
TOPP tests are registered in the CMakeLists.txt that I mentioned and each test consists of
- an executable call on a test dataset (by using either fixed command line parameters or an ini file)
- a FuzzyDiff call that compares the temporary output file of the last call and a reference test output that you have to provide
- a line to add a dependency of the FuzzyDiff call on the actual executable call (so they get executed after each other)

RobertDeibel commented:
Added experiment test file with commit de76582

timosachsenberg commented:
OK I think we kind of added to the confusion. Sorry.
The IDMapper::annotate that you extended can not be tested by a topp tool test because we can't save the mzML with mapped ids.
Best way is probably to use the IDMapper_test.cpp. In that test I would load an IdXML without RT information (just copy one of the input idXML for IDMapper and delete the RT="..." attributes or set them to -1) and map it to it's mzML file.
If the spectra have the id attached it means that mapping without RT information just based on native id worked.

RobertDeibel commented:
Added the Test. Let me know if it's alright to place the new code it in the same test that tests the old behaviour. 

 Also: I didn't really think and rebased this already published branch. I'll probably close this PR and open a new one, that does not include all the junk. Let me know if that is unnecessary.

timosachsenberg commented:
don't nee to open a new one. I will do a "squash and merge" later so it will be only one commit

timosachsenberg reviewed (approved):
looks good

timosachsenberg commented:
please "squash and merge"  upon approval
